crates/iota-indexer/src/apis/read_api.rs

```diff
@@ -172,15 +178,53 @@ impl ReadApiServer for ReadApi {
             );
         }
 
-        let mut futures = vec![];
-        for object_id in object_ids {
-            futures.push(self.get_object(object_id, options.clone()));
-        }
+        // Doesn't take care of missing objects.
+        let stored_objects = self
+            .inner
+            .multi_get_objects_in_blocking_task(object_ids.clone())
+            .await?;
 
-        futures::future::join_all(futures)
-            .await
+        // Map the returned `StoredObject`s to `ObjectID`
+        let object_map: HashMap<ObjectID, StoredObject> = stored_objects
             .into_iter()
-            .collect::<Result<Vec<_>, _>>()
+            .map(|obj| {
+                let object_id = ObjectID::try_from(obj.object_id.clone()).map_err(|_| {
+                    IndexerError::PersistentStorageDataCorruption(format!(
+                        "failed to parse ObjectID: {:?}",
+                        obj.object_id
+                    ))
+                })?;
+                Ok::<(ObjectID, StoredObject), IndexerError>((object_id, obj))
+            })
+            .collect::<Result<_, IndexerError>>()?;
+
+        let options = options.unwrap_or_default();
+        let resolver = self.inner.package_resolver();
+
+        // Create a future for each requested object id
+        let futures = object_ids.into_iter().map(|object_id| {
+            let options = options.clone();
+            let resolver = resolver.clone();
+            let maybe_stored = object_map.get(&object_id).cloned();
+            async move {
+                match maybe_stored {
+                    Some(stored) => {
+                        let object_read = stored.try_into_object_read(resolver).await?;
+                        self.object_read_to_object_response(object_read, options)
+                            .await
+                    }
+                    None => {
+                        self.object_read_to_object_response(
+                            ObjectRead::NotExists(object_id),
+                            options,
+                        )
+                        .await
+                    }
+                }
+            }
+        });
```

Comment:
Maybe something like this would be good.

```suggestion
        let futures = object_ids.into_iter().map(|object_id| {
            let options = options.clone();
            match object_map.get(&object_id) {
                Some(stored) => {
                    let resolver = resolver.clone();
                    let stored = stored.clone();
                    Box::pin(async move {
                        let object_read = stored.try_into_object_read(resolver).await?;
                        self.object_read_to_object_response(object_read, options)
                            .await
                    })
                }
                None => {
                    Box::pin(self.object_read_to_object_response(
                        ObjectRead::NotExists(object_id),
                        options,
                    ))
                }
            }
        });
```

You could also wrap the `HashMap` in an `Arc` to avoid cloning the objects.

Reply:
Hm the solution doesn't do it fully as of incompatible types. I wrapped the `HashMap` in an `Arc` to avoid the cloning the objects.